examples/passport-login/src/authentication-strategies/oauth2.ts

```diff
@@ -0,0 +1,123 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: example-passport-oauth2-login
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  AuthenticationStrategy,
+  UserIdentityService,
+  asAuthStrategy,
+  AuthenticationBindings,
+} from '@loopback/authentication';
+import {StrategyAdapter} from '@loopback/authentication-passport';
+import {Profile} from 'passport';
+import {Strategy, StrategyOptions} from 'passport-oauth2';
+import {Request, RedirectRoute} from '@loopback/rest';
+import {UserProfile, securityId} from '@loopback/security';
+import {User} from '../models';
+import {UserServiceBindings} from '../services';
+import {inject, bind, extensions, Getter} from '@loopback/core';
+
+export namespace PassportAuthenticationBindings {
+  export const OAUTH2_STRATEGY = 'passport.authentication.oauth2.strategy';
+}
+
+@bind(asAuthStrategy)
+export class Oauth2AuthStrategy implements AuthenticationStrategy {
+  name = 'oauth2';
+  passportstrategy: Strategy;
+  protected strategy: StrategyAdapter<User>;
+
+  /**
+   * create an oauth2 strategy
+   */
+  constructor(
+    /**
+     * enable extensions for provider specific oauth2 implementations
+     * reroute to the specific extension based on given provider name
+     */
+    @extensions(PassportAuthenticationBindings.OAUTH2_STRATEGY)
```

Comment:
Google and Facebook are registered as separate authentication strategies. Why does this  OAuth strategy have a getter to find facebook and google strategies?

Reply:
How do you intend to decorate a controller endpoint?

So far it is possible to say:

@authenticate('local')
or 
@authentication('facebook')
or
@authenticate('google')
or
@authenticate('oauth2') 

what will happen if it is 'oauth2'? Is it meant to handle many passports authentications that adhere to oauth2 settings or something?

Reply:
to avoid these many authenticate decorators and separate controllers for each provider, say google, facebook, twitter, etc etc, we have created a single controller with path param 'provider'

Reply:
@emonddr we use the extensions in the oauth2 strategy to handle specific providers